### PR TITLE
chore: prepare mainnet deployment

### LINKED ISSUE #46: Spore contract version management and release publish
# Description

We lack version management on Spore contract repository, we have to upload tags information to indicate our previous `v1` version and current `v2` version.

After deployment on CKB mainnet, we also have to publish deployment related information, such as `code_hash`, `tx_hash`, `out_index`, and corresponding contract names, etc.

# Tasks list
- [x] create tag for version v1
- [x] create tag for version v2
- [x] publish deployment information about testnet on README.md
- [x] publish deployment information about mainnet on README.md

### LINKED ISSUE #53: Different machine compiled different contract binary
# Description
different contract binary means different `code_hash` value, which we referred to define our version of contracts

### PR DESCRIPTION
**Pull request by ashuralyk**
# Description

after mainnet deployment, I complete `VERSIONS.md` doc that records our historical upload information with commit hash

# Tasks list
- [x] close #46 
- [x] close #53

### CODE REVIEW

**Comment by ashuralyk** [comment]
@ShookLyngs please see here to bump a new SDK version for mainnet adaption https://github.com/sporeprotocol/spore-contract/tree/chore/mainnet-deployment/deployment/migration/mainnet

File: docs/VERSIONS.md [hunk 1]
```diff
@@ -1 +1,51 @@
...
+**cluster_proxy**:
+|   |   |
+|---|---|
+|tx_hash|0x0231ea581bbc38965e10a2659da326ae840c038a9d0d6849f458b51d94870104|
+|index|0|
+|code_hash|0xbe8b9ce3d05a32c4bb26fe71cd5fc1407ce91e3a8b9e8719be2ab072cef1454b|
+
+## tag 0.2.1 (mainnet)
+
+commit hash: 40ef57a2d3689740bf0c6fa47938e52948252179
```
Comment (Flouse) [inline docs/VERSIONS.md:33]: Why is the commit of https://github.com/sporeprotocol/spore-contract/releases/tag/0.2.1 showing [53bfd4d4bbbee773ca10c1181982d9a27718826d](https://github.com/sporeprotocol/spore-contract/commit/53bfd4d4bbbee773ca10c1181982d9a27718826d)?

NOT 40ef57a2d3689740bf0c6fa47938e52948252179

image

**Review by Flouse — CHANGES_REQUESTED** [review]

File: docs/VERSIONS.md (hunk 1 above)
Comment (ashuralyk) [reply docs/VERSIONS.md:33]: because tag 0.2.1 created in a previous commit, and all of new commits will change repo's commit hash above then

Comment (Flouse) [reply docs/VERSIONS.md:33]: "Do NOT change a tag on a remote repository, unless sensitive data was leaked."

Comment (Flouse) [reply docs/VERSIONS.md:33]: ### See also:
`git help tag`
-> Read the DISCUSSION - On Re-tagging

Comment (ashuralyk) [reply docs/VERSIONS.md:33]: so what we gonna to do? or you can just do it by yourself, I have no idea for this.

Comment (Flouse) [reply docs/VERSIONS.md:33]: I am checking the diff between [53bfd4d4bbbee773ca10c1181982d9a27718826d](https://github.com/sporeprotocol/spore-contract/commit/53bfd4d4bbbee773ca10c1181982d9a27718826d) and 40ef57a2d3689740bf0c6fa47938e52948252179.

will give a suggestion later.

Comment (Flouse) [reply docs/VERSIONS.md:33]: Can't use [53bfd4d4bbbee773ca10c1181982d9a27718826d](https://github.com/sporeprotocol/spore-contract/commit/53bfd4d4bbbee773ca10c1181982d9a27718826d) to build:

```bash
capsule build --release -- --features release_export

  thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { inner: Error { inner: TomlError { message: "missing field `commit_hash`", original: Some("[[code_hash_list]]\n"), keys: ["code_hash_list"], span: Some(0..18) } } }', lib/build/src/lib.rs:111:29
```

I understood why you use https://github.com/sporeprotocol/spore-contract/commit/40ef57a2d3689740bf0c6fa47938e52948252179 to build.

## A simple solution
just create a new tag 0.2.2 for the mainnet release.

File: deployment/frozen/testnet.toml [hunk 2]
```diff
@@ -1,5 +1,11 @@
...
-commit_hash = "457e32d23b5f401cad837e52a568281784943e18"
 cluster = "598d793defef36e2eeba54a9b45130e4ca92822e1d193671f490950c3b856080"
 spore = "bbad126377d45f90a8ee120da988a2d7332c78ba8fd679aab478a19d6c133494"
+
+# version v0.2.1
+[[code_hash_list]]
+cluster = "7366a61534fa7c7e6225ecc0d828ea3b5366adec2b58206f2ee84995fe030075"
+spore = "5e063b4c0e7abeaa6a428df3b693521a3050934cf3b0ae97a800d1bc31449398"
+cluster_agent = "c986099b41d79ca1b2a56ce5874bcda8175440a17298ea5e2bbc3897736b8c21"
+cluster_proxy = "be8b9ce3d05a32c4bb26fe71cd5fc1407ce91e3a8b9e8719be2ab072cef1454b"
```
Comment (Flouse) [inline deployment/frozen/testnet.toml:11]:
```bash
# checkout commit 53bfd4d4bbbee773ca10c1181982d9a27718826d tag: 0.2.1
capsule clean
capsule build --release
```

And then I check the code_hashes:

## commit: 53bfd4d (capsule build --release)
- code_hash of contact spore              : 0x7e8bf78a62232caa2f5d47e691e8db1a90d05e93dc6828ad3cb935c01ec6d208
- code_hash of contact cluster            : 0x7366a61534fa7c7e6225ecc0d828ea3b5366adec2b58206f2ee84995fe030075
- code_hash of contact cluster_agent      : 0xc986099b41d79ca1b2a56ce5874bcda8175440a17298ea5e2bbc3897736b8c21
- code_hash of contact cluster_proxy      : 0xbe8b9ce3d05a32c4bb26fe71cd5fc1407ce91e3a8b9e8719be2ab072cef1454b
- code_hash of contact spore_extension_lua: 0x94a9b875911ace20f1f0d063a26495d14e4b04e32fd218261bb747f34e71ae47

### Could you check the code_hash of `contact spore` again?

File: deployment/frozen/mainnet.toml [hunk 3]
```diff
@@ -1 +1,4 @@
+# version v0.2.2
 [[code_hash_list]]
+cluster = "7366a61534fa7c7e6225ecc0d828ea3b5366adec2b58206f2ee84995fe030075"
+spore = "9cb02fa8621d79fff300b910d52665f594d3ad468142a66213cfaaca0409502c"
```
Comment (Flouse) [inline deployment/frozen/mainnet.toml:4]:
```bash
# checkout commit b562411fe3082d1362bae7a0d15c32ff94f80e0f tag: 0.2.2
capsule clean
capsule build --release -- --features release_export
```

And then I check the code_hashes:

## commit b562411fe3082d1362bae7a0d15c32ff94f80e0f (capsule build --release -- --features release_export)
- code_hash of contact spore              : 0x98b14c35863d82973b45fdc44c82dc04570dcde11c97f5c41f2b71c37dfa905f
- code_hash of contact cluster            : 0x7366a61534fa7c7e6225ecc0d828ea3b5366adec2b58206f2ee84995fe030075

### Could you check the code_hash of contact spore again?

Comment (ashuralyk) [reply deployment/frozen/mainnet.toml:4]: I know it, it caused by the new appending `frozen` version, so the Spore binary changed by supporting compatible `code_hash`

Comment (ashuralyk) [reply deployment/frozen/mainnet.toml]: @Flouse I have to change the linked commit hash of tag 0.2.2, so please change it again in your local machine

Comment (Flouse) [reply deployment/frozen/mainnet.toml]: Could you tell me exactly which commit you used to build and deploy to https://github.com/sporeprotocol/spore-contract/blob/f4d8f90a52fae3267442b3beaf14fda5c633b279/deployment/migration/mainnet/spore/2024-01-23-40ef57a2.json

Comment (ashuralyk) [reply deployment/frozen/mainnet.toml]: commit hash is not correct, I have to find out later

Comment (Flouse) [reply deployment/frozen/mainnet.toml]: > @Flouse I have to change the linked commit hash of tag 0.2.2, so please change it again in your local machine

Again, I suggest avoid changing tag, see https://github.com/sporeprotocol/spore-contract/pull/52#discussion_r1462794837 and https://git-scm.com/docs/git-tag#_on_re_tagging

You can simply create new tags such as 0.2.3, 0.2.4, if you need.

**Comment by Flouse** [comment]
## commit: f4d8f90

```
# Try to reproduce the code_hashes at commit: f4d8f90
capsule clean
capsule build --release -- --features release_export

ls -l build/release/*
```

    -rwxr-xr-x 1 ubuntu docker  46848 Jan 24 09:18 cluster
    -rwxr-xr-x 1 ubuntu docker  52960 Jan 24 09:18 cluster_agent
    -rwxr-xr-x 1 ubuntu docker  46192 Jan 24 09:18 cluster_proxy
    -rw-r--r-- 1 ubuntu docker 267432 Jan 24 09:17 libckblua.so
    -rwxr-xr-x 1 ubuntu docker  76752 Jan 24 09:18 spore
    -rwxr-xr-x 1 ubuntu docker  31424 Jan 24 09:17 spore_extension_lua

## commit: f4d8f90 (release)
```
- code_hash of contact spore              : 0x4a4dce1df3dffff7f8b2cd7dff7303df3b6150c9788cb75dcf6747247132b9f5
- code_hash of contact cluster            : 0x7366a61534fa7c7e6225ecc0d828ea3b5366adec2b58206f2ee84995fe030075
- code_hash of contact cluster_agent      : 0xcf79590446a6a526fe7ee2e64a0c5f216ae6755f79fb966fd03cd0e718157f69
- code_hash of contact cluster_proxy      : 0x2ed81372525ecef3da66da17152e2c48dfc2769443cc38577bce48df221625f1
- code_hash of contact spore_extension_lua: 0x94a9b875911ace20f1f0d063a26495d14e4b04e32fd218261bb747f34e71ae47
```

**Comment by ShookLyngs** [comment]
Code hashes from my local build (commit: https://github.com/sporeprotocol/spore-contract/commit/f4d8f90a52fae3267442b3beaf14fda5c633b279):
- Spore: 0x4a4dce1df3dffff7f8b2cd7dff7303df3b6150c9788cb75dcf6747247132b9f5
- Cluster: 0x7366a61534fa7c7e6225ecc0d828ea3b5366adec2b58206f2ee84995fe030075
- ClusterProxy: 0x2ed81372525ecef3da66da17152e2c48dfc2769443cc38577bce48df221625f1
- ClusterAgent: 0xcf79590446a6a526fe7ee2e64a0c5f216ae6755f79fb966fd03cd0e718157f69
- Mutant: 0x94a9b875911ace20f1f0d063a26495d14e4b04e32fd218261bb747f34e71ae47

**Comment by ashuralyk** [comment]
@ShookLyngs @Flouse it seems only Spore contract differs from yours, my `code_hash` of Spore contract is still this:

image

**Comment by ashuralyk** [comment]
@Flouse I have checked Spore binary size again in an Ubuntu server at the same commit hash, and I find out it really differs from mine (apple silicon M2), so it's a system problem, how do we fix it?

image

**Comment by Flouse** [comment]
> find out it really differs from mine (apple silicon M2), so it's a system problem, how do we fix it?

@ShookLyngs is also using macbook air with the m2 chip, got the same result as [mine](https://github.com/sporeprotocol/spore-contract/pull/52#issuecomment-1907725878).

Now, we know that the contract compilation is operating system independent. Only @ashuralyk 's dev environment produced a unexpected code hash of spore contract.

File: docs/VERSIONS.md [hunk 4]
```diff
@@ -1 +1,47 @@
...
+|code_hash|0xbe8b9ce3d05a32c4bb26fe71cd5fc1407ce91e3a8b9e8719be2ab072cef1454b|
+
+## tag 0.2.2 (mainnet)
+
+**spore**:
+| | |
+|-|-|
+|tx_hash|0xcb67e11a39594a183eae50664073a85fa0140ae25480df61114aca0e47726d38|
+|index|0|
+|code_hash|0x9cb02fa8621d79fff300b910d52665f594d3ad468142a66213cfaaca0409502c|
```
Comment (Flouse) [inline docs/VERSIONS.md:40]: maybe this code_hash should be deleted?
@ashuralyk

Comment (Flouse) [reply docs/VERSIONS.md:40]: **suggestions for next steps:**
1. spore-contract/pull/52
2. build and test
3. create a tag on master branch  for mainnet
4. check the codehash, do the redeployment on mainnet
5. create a new PR to update the deployment info, such as `VERSIONs.md`

Comment (Flouse) [inline docs/VERSIONS.md:40]:
```suggestion
```

**Review by Flouse — CHANGES_REQUESTED** [review]

**Review by Flouse — APPROVED** [review]

**Review by ShookLyngs — APPROVED** [review]

**Pull request merged**